nexus/src/db/queries/external_ip.rs

```diff
@@ -437,18 +442,21 @@ impl NextExternalIp {
         out.push_sql(") AS candidate_ip FROM ");
         IP_POOL_RANGE_FROM_CLAUSE.walk_ast(out.reborrow())?;
         out.push_sql(" WHERE ");
-        if let Some(ref pool_id) = self.ip.pool_id() {
-            out.push_identifier(dsl::ip_pool_id::NAME)?;
-            out.push_sql(" = ");
-            out.push_bind_param::<sql_types::Uuid, Uuid>(pool_id)?;
-        } else {
-            out.push_sql("(");
-            out.push_identifier(dsl::project_id::NAME)?;
-            out.push_sql(" = ");
-            out.push_bind_param::<sql_types::Uuid, Uuid>(self.ip.project_id())?;
-            out.push_sql(" OR ");
-            out.push_identifier(dsl::project_id::NAME)?;
-            out.push_sql(" IS NULL)");
+        match self.ip.source() {
```

Comment:
This is the callsite which motivated the creation of `IpSource`. Without that structure, we'd have an `Option<pool_id>` and an `Option<project_id>`. Now that the `project_id` is optional, *both* could be null. In that case, this query would be invalid.

Reply:
I think I understand your point. For instance-related IP addresses, we always have some project, and possibly a pool. For service-related IP addresses we never have a project, and always have a pool. (I think this just restates the invariants maintained by the constructors like `InstanceExternalIp::for_floating` etc.)

I feels a bit weird to me to have two fields though, the project and this new source field, because they're sometimes duplicates. What if we combine the pool and project (and maybe the kind) into a single field that captures these? Something like this, though the name is not great now:

```rust
pub enum IpSource {
    Instance { project_id: Uuid, pool_id: Option<Uuid> },
    Service { pool_id: Uuid },
}
```

I say "maybe the kind" because we now have some duplication there too. It might be better to just have this enum with a variant for each kind, where the instance-related ones all have the same fields. I dunno, WDYT?

Reply:
I like this proposal. I had to shuffle around some of the query construction, but the structures make more sense when specified within the enum that you described.

Updated!